### PR TITLE
Add `BlockDiagonalSolver` for combining `LinearSolver`s

### PR DESCRIPTION
**Pull request by davschneller**
This PR introduces a new class `BlockDiagonalSolver`. The features are:
* Applies several `LinearSolvers` to parts of a `BlockVector`.
* Works as a `LinearSolver` on `BlockVectorSpace`s. (it also inherits from `LinearSolver`)
* Includes tests.
* Updates projector code to use `KroneckerLinearSolver` and the introduced `BlockDiagonalSolver`.

### CODE REVIEW

File: psydac/linalg/tests/test_block.py [hunk 1]
```diff
@@ -92,6 +98,100 @@ def test_block_linear_operator_serial_dot( n1, n2, p1, p2, P1, P2  ):
...
 
+#===============================================================================
+@pytest.mark.parametrize( 'n1', [8,16] )
+@pytest.mark.parametrize( 'n2', [8,12] )
+@pytest.mark.parametrize( 'p1', [1,2,3] )
+@pytest.mark.parametrize( 'p2', [1,2,3] )
+@pytest.mark.parametrize( 'P1', [True, False] )
+@pytest.mark.parametrize( 'P2', [True, False] )
+def test_block_diagonal_solver_serial_dot( n1, n2, p1, p2, P1, P2  ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:109]: Typo: reproducability -> reproducibility

File: psydac/linalg/tests/test_block.py [hunk 2]
```diff
@@ -101,6 +201,8 @@ def test_block_linear_operator_serial_dot( n1, n2, p1, p2, P1, P2  ):
 @pytest.mark.parametrize( 'P2', [True, False] )
 
 def test_block_matrix( n1, n2, p1, p2, P1, P2  ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:204]: Typo

File: psydac/linalg/tests/test_block.py [hunk 3]
```diff
@@ -180,6 +282,8 @@ def test_block_matrix( n1, n2, p1, p2, P1, P2  ):
 @pytest.mark.parametrize( 'P2', [True, False] )
 
 def test_block_2d_array_to_stencil_1( n1, n2, p1, p2, P1, P2 ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:285]: Typo

File: psydac/linalg/tests/test_block.py [hunk 4]
```diff
@@ -210,6 +314,8 @@ def test_block_2d_array_to_stencil_1( n1, n2, p1, p2, P1, P2 ):
 @pytest.mark.parametrize( 'P2', [True, False] )
 
 def test_block_2d_array_to_stencil_2( n1, n2, p1, p2, P1, P2 ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:317]: Typo

File: psydac/linalg/tests/test_block.py [hunk 5]
```diff
@@ -246,6 +352,8 @@ def test_block_2d_array_to_stencil_2( n1, n2, p1, p2, P1, P2 ):
 @pytest.mark.parallel
 
 def test_block_linear_operator_parallel_dot( n1, n2, p1, p2, P1, P2, reorder ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:355]: Typo

File: psydac/linalg/tests/test_block.py [hunk 6]
```diff
@@ -316,6 +424,118 @@ def test_block_linear_operator_parallel_dot( n1, n2, p1, p2, P1, P2, reorder ):
...
+@pytest.mark.parametrize( 'n1', [8,16] )
+@pytest.mark.parametrize( 'n2', [8,32] )
+@pytest.mark.parametrize( 'p1', [1,3] )
+@pytest.mark.parametrize( 'p2', [1,2] )
+@pytest.mark.parametrize( 'P1', [True, False] )
+@pytest.mark.parametrize( 'P2', [True, False] )
+@pytest.mark.parametrize( 'reorder', [True, False] )
+@pytest.mark.parallel
+def test_block_diagonal_solver_parallel_dot( n1, n2, p1, p2, P1, P2, reorder  ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:437]: Typo

File: psydac/linalg/tests/test_block.py [hunk 7]
```diff
@@ -20,6 +23,8 @@
 @pytest.mark.parametrize( 'P2', [True, False] )
 
 def test_block_linear_operator_serial_dot( n1, n2, p1, p2, P1, P2  ):
+    # set seed for reproducability
```
Comment (yguclu) [inline psydac/linalg/tests/test_block.py:26]: Typo: reproducability -> reproducibility

**Review by yguclu — CHANGES_REQUESTED** [review]
Thanks @davschneller, this is very neat. I have a couple of comments

File: psydac/linalg/block.py [hunk 8]
```diff
@@ -645,3 +645,105 @@ def topetsc( self ):
...
+    def space( self ):
+        return self._space
+
+    # ...
+    def solve( self, rhs, out=None, transposed=False ):
+        assert isinstance(rhs, BlockVector)
+        assert rhs.space is self.space
+        if out is not None:
+            assert isinstance(out, BlockVector)
+            out *= 0.0
```
[the remark this answers is not in the document]
Comment (davschneller) [reply psydac/linalg/block.py]: Yes, that should be in there; thanks for the notice; it has been changed with the latest commit. Tests and documentation have been updated as well to include the `out is rhs` and `out is not None and out is not rhs` case.

**Review by yguclu — APPROVED** [review]
Very good! Thanks for the changes

**Review by saidctb — APPROVED** [review]

**Pull request merged**